ios/RCTWebRTC/ScreenCapturer.m

```diff
@@ -0,0 +1,227 @@
+    
+    NSData *messageData = CFBridgingRelease(CFHTTPMessageCopyBody(_framedMessage));
+    
+    // Copy the pixel buffer
+    CVReturn status = CVPixelBufferCreate(kCFAllocatorDefault, width, height, kCVPixelFormatType_32BGRA, NULL, &_imageBuffer);
+    if (status != kCVReturnSuccess) {
+        NSLog(@"CVPixelBufferCreate failed");
+        return false;
+    }
+    
+    [self copyImageData:messageData toPixelBuffer:&_imageBuffer];
+    if (_imageBuffer == nil) {
+        return false;
+    }
+
+    return true;
+}
+
+- (void)copyImageData:(NSData *)data toPixelBuffer:(CVPixelBufferRef*)pixelBuffer {
+    CVPixelBufferLockBaseAddress(*pixelBuffer, 0);
+        
+    CIImage *image = [CIImage imageWithData:data];
+    [self.imageContext render:image toCVPixelBuffer:*pixelBuffer];
+
+    CVPixelBufferUnlockBaseAddress(*pixelBuffer, 0);
+}
+
+@end
+
+// MARK: -
+
+@interface ScreenCapturer () <NSStreamDelegate>
+
+@property (nonatomic, strong) SocketConnection *connection;
+@property (nonatomic, strong) Message *message;
+@property (nonatomic, strong) NSDate *startDate;
+
+@end
+
+@implementation ScreenCapturer {
+    NSInteger _readLength;
+}
+
+- (instancetype)initWithDelegate:(__weak id<RTCVideoCapturerDelegate>)delegate {
+    return [super initWithDelegate:delegate];
+}
+
+- (void)startCaptureWithConnection:(SocketConnection *)connection {
+    self.connection = connection;
+    self.message = nil;
+    self.startDate = nil;
+    
+    [self.connection openWithStreamDelegate:self];
+}
+
+- (void)stopCapture {
+    [self.connection close];
+}
+
+// MARK: Private Methods
+
+- (void)readBytesFromStream:(NSInputStream *)stream {
+    if (!stream.hasBytesAvailable) {
+        return;
+    }
+    
+    const NSUInteger maxReadLength = kConnectionBufferSize;
+    
+    if (!self.message) {
+        self.message = [[Message alloc] init];
+        _readLength = maxReadLength;
+        
+        __weak __typeof__(self) weakSelf = self;
+        self.message.didComplete = ^(BOOL success, Message *message) {
+            if (success) {
+                [weakSelf didCaptureVideoFrame:message.imageBuffer];
+            }
+            
+            weakSelf.message = nil;
+        };
+    }
+    
+    uint8_t buffer[_readLength];
+    NSInteger numberOfBytesRead = [stream read:buffer maxLength:_readLength];
+    if (numberOfBytesRead < 0) {
+        NSLog(@"error reading bytes from stream");
+        return;
+    }
+    
+    _readLength = [self.message appendBytes:buffer length:numberOfBytesRead];
+    if (_readLength == -1 || _readLength > maxReadLength) {
+        _readLength = maxReadLength;
+    }
+}
+
+- (void)didCaptureVideoFrame:(CVPixelBufferRef)pixelBuffer {
+    if (!self.startDate) {
+        self.startDate = [NSDate date];
+    }
+        
+    RTCCVPixelBuffer *rtcPixelBuffer = [[RTCCVPixelBuffer alloc] initWithPixelBuffer: pixelBuffer];
+    int64_t timeStampNs = fabs([self.startDate timeIntervalSinceNow]) * kNanosecondsPerSecond;
```

Comment:
Isn't NSDate too "heavy" for this? I tnow it doesn't directly translate, but this what the WebRTC sample does:

`int64_t timeStampNs =
      CMTimeGetSeconds(CMSampleBufferGetPresentationTimeStamp(sampleBuffer)) * NSEC_PER_SEC;`

Reply:
The WebRTC code is just reading the time stamp from the CMSampleBufferRef, which we don't have. It's either we compute it on the client side and send it as part of the message, or compute it here. I chose the latest.

Reply:
Fixed. I replaced the implementation to use `mach_time` instead of `NSDate`